docs/decisions/0025-stripe-event-consumption-delivery.rst

```diff
@@ -0,0 +1,108 @@
+0025 Stripe Event Consumption And Delivery
+******************************************
+
+Status
+======
+**In progress** (March 2025)
+
+Context
+=======
+
+We intend to utilize Stripe to manage subscription and payment lifecycles for the Enterprise Self-Service Purchasing
+feature. We will need a robust way to ingest events from Stripe, handle those events, and forward them to SalesForce.
+For example, when stripe emits an ``invoice.paid`` event, that needs to be communicated to SalesForce in order to
+trigger the creation/update of Customers, Opportunities, etc. Furthermore, that same event needs to be communicated to
+our edX Enterprise backend so that we can immediately followup with the customer to confirm that we have begun to
+provision their product.
+
+At minimum, "robust" means:
+- Spoofed events are discarded.
+- Network failure does not cause dropped events.
+- Network failure does not cause events to be handled multiple times, or duplicate products provisioned.
+
+Decision
+========
+
+The selected solution can be described as having three decoupled parts:
+
+#. **Ingest:** Implement a new webhook endpoint in the edX Enterprise backend.
+
+   * IDA and URL: TBD.
+
+   * The view logic should robustly ingest Stripe events, storing them in the DB.
+
+     * Use `signature validation <https://docs.stripe.com/webhooks#verify-official-libraries>`_.
+
+     * Use `de-duplication <https://docs.stripe.com/webhooks#handle-duplicate-events>`_.
+
+   * After successful ingestion, respond with HTTP 200 to tell Stripe to stop retrying sending the event.
+
+#. **Pick:** Implement an asynchronous event picker to continuously try to pick events to handle/forward.
+
+   * Architecturally, this is an infinite loop via management command running in a dedicated container/pod.
+
+     * Alternatively, this might work as a Celery "beat" scheduled for every 10 seconds.  This could reduce
```

Comment:
[question] How would this alternative 10-second "beat" schedule compare to the infinite loop approach in terms of event delivery timeliness? Just pondering the tradeoff between potentially slower delivery and reduced code/infra.

Reply:
Ah, good find on the additional downsides. A 1-second sleep with the infinite loop seems reasonable!